indexer/indexer.py

```diff
@@ -0,0 +1,337 @@
+        # Set all kwargs as instance attributes
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    def index(self, bundle_uuid, bundle_version, *args, **kwargs):
+        """
+        Indexes the data files.
+
+        Triggers the actual indexing process
+
+        :param bundle_uuid: The bundle_uuid of the bundle that will be indexed.
+        :param bundle_version: The bundle of the version.
+        """
+        raise NotImplementedError(
+            'users must define index to use this base class')
+
+    def special_fields(self, *args, **kwargs):
+        """
+        Add any special fields that may be missing.
+
+        Gets any special field that may not be available directly from the
+        metadata.
+
+        :return: a dictionary of all the special fields to be added.
+        """
+        raise NotImplementedError(
+            'users must define special_fields to use this base class')
+
+    def merge(self, doc_contents, **kwargs):
+        """
+        Merge the document with the contents in ElasticSearch.
+
+        merge() should take the results of get_item() and harmonize it
+        with whatever is present in ElasticSearch to avoid blind overwritting
+        of documents. Users should implement their own protocol.
+
+        :param doc_contents: Current document to be indexed.
+        :return: The harmonized document which can overwrite existing entry.
+        """
+        raise NotImplementedError(
+            'users must define merge to use this base class')
+
+    def load_doc(self, doc_uuid, doc_contents, * args, **kwargs):
+        """
+        Load a document into ElasticSearch.
+
+        Load 'doc_contents' as a document in ElasticSearch (ES) with a
+        uuid 'doc_uuid'.
+
+        :param index_name: The name of the index to load into ElasticSearch.
+        :param doc_type: The name of the document type in ElasticSearch.
+        :param doc_uuid: The uuid which will root the ElasticSearch document.
+        :param doc_contents: The contents that will be loaded into ES.
+        :return:
+        """
+        # Loads the index document file into elasticsearch
+        self.es_client.index(index=self.index_name,
+                             doc_type=self.doc_type,
+                             id=doc_uuid,
+                             body=doc_contents)
+
+    def create_mapping(self, *args, **kwargs):
+        """
+        Create the ElasticSearch mapping.
+
+        To be overwritten by the class inheriting from here.
+        This class should return the dictionary corresponding to the mapping
+        to be loaded into ElasticSearch. It should use the index_mapping_config
+        attribute as appropriate.
+
+        :return: Dictionary describing the ElaticSearch mapping.
+        """
+        raise NotImplementedError(
+            'users must define create_mapping to use this base class')
+
+    def load_mapping(self, *args, **kwargs):
```

Comment:
This should be called `store_mapping` or `save_mapping`. In my book loading sounds line reading, this method writes.